### PR TITLE
Add singularity inspect metalad extractor

### PR DESCRIPTION
**Pull request by asmacdo**
Fixes: https://github.com/datalad/datalad-container/issues/198

This PR is the result of Datalad Metalad hackathon last week!

TODOs
- [x] add some rudimentary unittest
- [x] make CI happy

### CODE REVIEW

File: changelog.d/20230307_164111_austin_add_metalad_extractor.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+### 🚀 Enhancements and New Features
+
+- Add metalad extractor using `singularity inspect`.
+  Fixes https://github.com/datalad/datalad-container/issues/198 via
+  https://github.com/datalad/datalad-next/pull/200 (by @asmacdo )
```
Comment (asmacdo) [inline changelog.d/20230307_164111_austin_add_metalad_extractor.md:5]:
```suggestion
 https://github.com/datalad/datalad-container/pull/200 (by @asmacdo )
```

**Review by yarikoptic — CHANGES_REQUESTED** [review]
Also will need a rudimentary unittest on extracting metadata from a sample container. There is already some which use singularity etc, so should be similar I guess.  Test should skip if no metadata or singularity available.  But metalad should be installed in CI to actually do run that test.

File: requirements-devel.txt [hunk 2]
```diff
@@ -1,2 +1,3 @@
 # requirements for a development environment
 -e .[devel]
+scriv
```
Comment (yarikoptic) [inline requirements-devel.txt:3]: this inspired me for https://github.com/datalad/datalad-extension-template/pull/56 . Then here it would be `-e .[devel,devel-utils]` if so really needed. But since this has nothing to do with adding metadata extractor -- shouldn't even be in this PR. Feel welcome to submit a separate PR with such changes.

**Comment by yarikoptic** [comment]
re appveyor -- I did that fix in https://github.com/datalad/datalad-container/pull/190 already which I merged today. Just rebase your branch on current master and it should be gone.

File: datalad_container/extractors/metalad_container.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
```
[the remark this answers is not in the document]
Comment (asmacdo) [reply datalad_container/extractors/metalad_container.py:6]: IIUC it wasn't necessary (it worked) because it was being treated as an [Implicit Namespace](https://peps.python.org/pep-0420/)

But since that wasn't deliberate, I've added an __init__.py

File: datalad_container/extractors/metalad_container.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+from uuid import UUID
+
+CURRENT_VERSION = "0.0.1"
+
+
+try:
+    from datalad_metalad.extractors.base import DataOutputCategory, ExtractorResult, FileMetadataExtractor
+    from datalad_metalad import get_file_id
+except Exception as e:
+    raise
```
[the remark this answers is not in the document]
Comment (asmacdo) [reply datalad_container/extractors/metalad_container.py:23]: this is a placeholder for https://github.com/datalad/datalad-container/pull/200/files#r1129521291

File: datalad_container/extractors/metalad_container.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+            })
+
+    def _singularity_inspect(self, path) -> str:
+        data = subprocess.run(
+            ["singularity", "inspect", "--json", path],
+            check=True,
+            stdout=subprocess.PIPE).stdout.decode()
+        return json.loads(data)
+
+    def _singularity_version(self) -> str:
```
[the remark this answers is not in the document]
Comment (asmacdo) [reply datalad_container/extractors/metalad_container.py:72]: Agreed its low risk, but IMO its also low reward. `singularity version` seems basically free in comparison to `inspect`?

File: setup.cfg [hunk 6]
```diff
@@ -41,6 +41,9 @@ datalad.extensions =
     # valid datalad interface specification (see demo in this extensions)
     container = datalad_container:command_suite
 
+datalad.metadata.extractors =
+    singularity_inspect = datalad_container.extractors.metalad_container:MetaladSingularityInspect
```
Comment (asmacdo) [inline setup.cfg:45]: I changed this to `singularity_inspect` since `metalad` seemed well implied.

File: datalad_container/extractors/metalad_container.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (asmacdo) [reply datalad_container/extractors/metalad_container.py:71]: Given that there is one instance of the class per file, (bc parent is FileMetadataExtractor), I don't think we can use the `@property` pattern. 

Looked over the [External Versions]( https://github.com/datalad/datalad/blob/8d431b654337202bd6856c08240c22e370b410dd/datalad/support/external_versions.py#L153-L164) Cool, we have something very similar in ReproMan (...probably exactly this  :) This is a cool pattern, but IMO not worth the time for this particular performance bump.

However, this does seem like a cool and reasonable feature request for metalads `BaseMetadataExtractor`. I'm thinking `def _get_cacheable_metadata`. Shall I file an issue of there?

File: setup.cfg [hunk 7]
```diff
@@ -41,6 +41,9 @@ datalad.extensions =
     # valid datalad interface specification (see demo in this extensions)
     container = datalad_container:command_suite
 
+datalad.metadata.extractors =
+    metalad_container = datalad_container.extractors.metalad_container:MetaladContainer
```
[the remark this answers is not in the document]
Comment (asmacdo) [reply setup.cfg]: Easier than expected because this endpoint is under `datalad.metalad`

Without `metalad` installed:

``` 
datalad meta-extract -d . singularity_inspect images/bids/bids-pymvpa--1.0.2.sing  | jq
datalad: Unknown command 'meta-extract'.  See 'datalad --help'.

Did you mean any of these?
        metadata
```

With `metalad` installed.

```
{
  "type": "file",
  "dataset_id": "b02e63c2-62c1-11e9-82b0-52540040489c",
  "dataset_version": "9ed0a39406e518f0309bb665a99b64dec719fb08",
  "path": "images/bids/bids-pymvpa--1.0.2.sing",
  "extractor_name": "singularity_inspect",
  "extractor_version": "0.0.1",
  "extraction_parameter": {},
  "extraction_time": 1678382512.413365,
  "agent_name": "Austin Macdonald",
  "agent_email": "austin@dartmouth.edu",
  "extracted_metadata": {
    "@id": "datalad:SHA1-s993116191--cc7ac6e6a31e9ac131035a88f699dfcca785b844",
    "type": "file",
    "path": "images/bids/bids-pymvpa--1.0.2.sing",
    "content_byte_size": 0,
    "comment": "SingularityInspect extractor executed at 1678382512.2972443",
    "singularity_version": "apptainer version 1.1.5-1.fc37",
    "singularity_inspect": {
      "data": {
        "attributes": {
          "labels": {
            "org.label-schema.build-date": "Thu,_19_Dec_2019_14:58:41_+0000",
            "org.label-schema.build-size": "2442MB",
            "org.label-schema.schema-version": "1.0",
            "org.label-schema.usage.singularity.deffile": "Singularity.bids-pymvpa--1.0.2",
            "org.label-schema.usage.singularity.deffile.bootstrap": "docker",
            "org.label-schema.usage.singularity.deffile.from": "bids/pymvpa:v1.0.2",
            "org.label-schema.usage.singularity.version": "2.5.2-feature-squashbuild-secbuild-2.5.6e68f9725"
          }
        }
      },
      "type": "container"
    }
  }
}
```

File: datalad_container/extractors/metalad_container.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (yarikoptic) [reply datalad_container/extractors/metalad_container.py]: sorry -- this no longer leads anywhere specifically seems to me. So what is it for? may be just a comment instead of try/except would suffice for a placholder?

File: setup.cfg (hunk 7 above)
[the remark this answers is not in the document]
Comment (yarikoptic) [reply setup.cfg]: coolio then

File: datalad_container/extractors/metalad_container.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""MetadataRecord extractor for files stored in Datalad's own core storage"""
```
Comment (yarikoptic) [inline datalad_container/extractors/metalad_container.py:9]: adjust docstring and filename also to reflect that it is not a generic  extractor for any container

File: datalad_container/extractors/metalad_container.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (yarikoptic) [reply datalad_container/extractors/metalad_container.py:71]: the easiest way to implement is just `@functools.lru_cache(1)`  for the function, no need for any special handling/code IMHO.

File: datalad_container/extractors/metalad_container.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (asmacdo) [reply datalad_container/extractors/metalad_container.py]: The try/except was where I was planning on gating the plugin in case metalad isn't installed. 

Since it only has the 1 entry point, and if metalad is not installed the entry point fails before we get to this code, I dont think we need to do anything fancy with the imports.

**Comment by yarikoptic** [comment]
re test
- let's use image from `tests/test_run.py:testimg_url = 'shub://datalad/datalad-container:testhelper'`
- create a session scope fixture. More advanced example is in dandi-cli https://github.com/dandi/dandi-cli/blob/HEAD/dandi/tests/fixtures.py#L342 but in your case it would be simpler -- it would `singularity pull` that image into temporary location, and give that path as what it yields as a "fixture" to underlying test. Then multiple tests could reuse that download.
- for running external commands use DataLad's `WitlessRunner` , smth like `out = WitlessRunner().run(cmd, protocol=StdOutCapture)` but see other uses in the tests of datalad.

**Comment by yarikoptic** [comment]
re external_versions -- just "patch" it to provide function to load version for singularity and apptainer. Here is how

```
In [8]: external_versions.CUSTOM['cmd:singularity'] = lambda : '0.1.1'

In [9]: external_versions['cmd:singularity']
Out[9]: LooseVersion ('0.1.1')
```
so replace lambda with yours. For that probably just do in  `from datalad.support.external_versions import external_versions` and do that patching. This would make `external_versions["cmd:singularity"]` available throughout for all uses of `external_versions` in that process, including I think reporting in `datalad wtf` output as long as datalad_container is imported (via `datalad.api` or elsehow)

**Comment by asmacdo** [comment]
[  ] skipif test if datalad metalad not installed
[ ] install metalad for travis

File: datalad_container/extractors/metalad_container.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
```
Comment (asmacdo) [inline datalad_container/extractors/metalad_container.py:2]: @yarikoptic is there a reason we set noexpand tab? Seems wrong for python.

Comment (yarikoptic) [reply datalad_container/extractors/metalad_container.py:2]: yeah, I think this some historical artifact. Feel welcome to fix everywhere (in separate PR) in all extensions . I think we did similar in datalad core repo but even there not everywhere

```shell
❯ git grep -h '^# ex:' | sort | uniq -c
    264 # ex: set sts=4 ts=4 sw=4 et:
     14 # ex: set sts=4 ts=4 sw=4 noet:
```

so here is a little good menial contribution to make ;)

File: datalad_container/extractors/tests/test_metalad_container.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+            'org.label-schema.build-size',
+            'org.label-schema.usage.singularity.deffile',
+            'org.label-schema.usage.singularity.deffile.from',
+            'org.label-schema.usage.singularity.deffile.bootstrap',
+            'org.label-schema.usage.singularity.version',
+        ]
+        for label in assert_in_labels:
+            assert label in assert_in_labels
+        assert labels['org.label-schema.schema-version'] == '1.0'
+        assert result['type'] == 'container'
```
Comment (yarikoptic) [inline datalad_container/extractors/tests/test_metalad_container.py:62]: add a test which actually does full metadata extraction invocation -- look how metalad does in https://github.com/yarikoptic/datalad-metalad/blob/HEAD/datalad_metalad/tests/test_extract.py testing of `meta_extract`. Also since we test on specific image -- test more - e.g. specific values (size, date), may be just a full value 
```
    "container_inspect": {
        "data": {
          "attributes": {
            "labels": {
              "org.label-schema.build-date": "Thu,_19_Dec_2019_14:58:41_+0000",
              "org.label-schema.build-size": "2442MB",
              "org.label-schema.schema-version": "1.0",
              "org.label-schema.usage.singularity.deffile": "Singularity.bids-pymvpa--1.0.2",
              "org.label-schema.usage.singularity.deffile.bootstrap": "docker",
              "org.label-schema.usage.singularity.deffile.from": "bids/pymvpa:v1.0.2",
              "org.label-schema.usage.singularity.version": "2.5.2-feature-squashbuild-secbuild-2.5.6e68f9725"
            }
          }
        },
        "type": "container"
      }
```

**Review by yarikoptic — CHANGES_REQUESTED** [review]

File: datalad_container/extractors/tests/test_metalad_container.py [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    raise SkipTest("skipping metalad tests")
+
+# Must come after skiptest or imports will not work
+from datalad_container.extractors.metalad_container import MetaladSingularityInspect
+
+
+class TestMetaladSingularityInspect:
+
+    @with_tempfile
+    def test__singularity_inspect_nofile(self, path=None):
```
[the remark this answers is not in the document]
Comment (yarikoptic) [reply datalad_container/extractors/tests/test_metalad_container.py:39]: ah, ok, makes sense - so I am ok now with it as is

File: datalad_container/extractors/tests/test_metalad_container.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+)
+
+if not external_versions["datalad_metalad"]:
+    raise SkipTest("skipping metalad tests")
+
+# Must come after skiptest or imports will not work
+from datalad_container.extractors.metalad_container import MetaladSingularityInspect
+
+
+class TestMetaladSingularityInspect:
```
[the remark this answers is not in the document]
Comment (asmacdo) [reply datalad_container/extractors/tests/test_metalad_container.py:36]: :+1: 

FYI: I put it in a class to think ahead, assuming that the container extractor will have more functionality someday, also tested in this file. The organization I've used in the past is a putting all of the tests for a function (ie singularity inspect) inside a class, and having a separate class for each function. Given that this is it for now, this could be premature.

File: datalad_container/extractors/_load_singularity_versions.py [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+from datalad.support.external_versions import external_versions
+
+
+def __get_apptainer_version():
+    try:
+        out = WitlessRunner().run("apptainer --version", protocol=StdOutCapture)
+        version = out['stdout'].strip()
+
+    except FileNotFoundError as e:
+        return None
```
[the remark this answers is not in the document]
Comment (asmacdo) [reply datalad_container/extractors/_load_singularity_versions.py]: I was able to trim this up nicely, but not able to fully go with this suggestion, I still have to strip "apptainer " from the string, and for singularity, I have to make sure that apptainer version is not the same.

File: datalad_container/extractors/_load_singularity_versions.py [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+    StdOutCapture,
+    WitlessRunner,
+)
+from datalad.support.external_versions import external_versions
+
+
+def __get_apptainer_version():
+    version = WitlessRunner().run("apptainer --version", protocol=StdOutCapture)['stdout'].strip()
+    strip_str = len("apptainer version ")
+    return version[strip_str:]
```
Comment (yarikoptic) [inline datalad_container/extractors/_load_singularity_versions.py:20]: another pattern for the same thing for your consideration

```suggestion
    return version.split("apptainer version", 1)[1].strip()
```

File: datalad_container/extractors/_load_singularity_versions.py [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+
+
+def __get_singularity_version():
+    version = WitlessRunner().run("singularity version", protocol=StdOutCapture)['stdout'].strip()
+
+    # It may be possible to have both apptainer and singularity installed.
+    # If singularity is installed independently, the versions will not match.
```
Comment (yarikoptic) [inline datalad_container/extractors/_load_singularity_versions.py:27]: is that guaranteed -- they really cannot be the same?

Comment (asmacdo) [reply datalad_container/extractors/_load_singularity_versions.py:27]: Now that you mention it, I think it's ok to have cmd: singularity return the apptainer version.

It wouldn't be ideal as the code currently exists because it collects both singularity and apptainer version info, but once I use the dynamic command as you suggested, it's a moot point.

+1 I'm convinced

File: changelog.d/20230307_164111_austin_add_metalad_extractor.md [hunk 16]
```diff
@@ -0,0 +1,5 @@
+### 🚀 Enhancements and New Features
+
+- Add metalad extractor using `singularity inspect`.
```
Comment (yarikoptic) [inline changelog.d/20230307_164111_austin_add_metalad_extractor.md:3]: not liking name 100% but due to singlularity/apptainer duality let's keep as is and do mention in changelog
```suggestion
- Add metalad extractor `container_inspect` using `singularity` or `apptainer` `inspect`.
```

File: datalad_container/extractors/tests/test_metalad_container.py [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+                    'org.label-schema.usage.singularity.deffile.from': 'debian:stable-slim', 'org.label-schema.usage.singularity.version':
+                    '2.5.0-feature-squashbuild-secbuild-2.5.0.gddf62fb5'
+                }
+            }
+        },
+        'type': 'container'
+    }
+    assert result == expected_result
+
+def test_extract(singularity_test_image):
```
Comment (yarikoptic) [inline datalad_container/extractors/tests/test_metalad_container.py:67]: should be skipped if no singularity or apptainer installed.

File: datalad_container/extractors/tests/test_metalad_container.py [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+)
+
+if not external_versions["datalad_metalad"]:
+    raise SkipTest("skipping metalad tests")
+
+# Must come after skiptest or imports will not work
+from datalad_container.extractors.metalad_container import MetaladContainerInspect
+
+
+@with_tempfile
```
Comment (yarikoptic) [inline datalad_container/extractors/tests/test_metalad_container.py:37]: next two tests rely on `apptainer` to be present . ATM AFAIK our CI installs singularity... could we make it more flexible -- e.g. add some module level const "CONTAINER_SYSTEM", skip entire module if None available, and store  string either `"singularity"` or `"apptainer"` and then use it in this tests, thus avoiding hardcoding any and allowing for either to be used here.

that would also avoid possible shadowing of CalledProcessError due to simply not finding that executable instead of it unable to work with  bad path

File: datalad_container/extractors/tests/test_metalad_container.py [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+def test__container_inspect_nofile(path=None):
+    """Singularity causes CalledProcessErorr if path DNE."""
```
Comment (yarikoptic) [inline datalad_container/extractors/tests/test_metalad_container.py:39]:
```suggestion
    """Singularity causes CalledProcessError if path DNE."""
```

**Review by yarikoptic — CHANGES_REQUESTED** [review]
The main point is to make tests "execute" (not just skip) and pass on CI

File: datalad_container/extractors/tests/test_metalad_container.py [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+    }
+    assert result == expected_result
+
+def test_extract(singularity_test_image):
+    ds = singularity_test_image["ds"]
+    path = singularity_test_image["img_path"]
+    result = meta_extract(dataset=ds, extractorname="container_inspect", path=path)
+    assert len(result) == 1
+
+    assert "extracted_metadata" in result[0]["metadata_record"]
```
Comment (yarikoptic) [inline datalad_container/extractors/tests/test_metalad_container.py:80]:
```suggestion
    assert  result[0]["metadata_record"]["extracted_metadata"]
```
so we know that at least some content is assigned and not only key defined

File: datalad_container/tests/fixtures/singularity_image.py [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+        url=TEST_IMG_URL,
+        image=fixture_file_name,
+    )
+    img_path = ds.pathobj / fixture_file_name
+    ds.get(img_path)
+    return {"ds": ds, "img_path": img_path}
+
+@pytest.fixture(scope="session")
+def container_command():
+    """Not a very useful function other than to add session scope."""
```
Comment (yarikoptic) [inline datalad_container/tests/fixtures/singularity_image.py:28]: could have just made that function "lazy" via using e.g. `@functools.cache` but ok for now
